doc/cephadm/services/snmp-gateway.rst

```diff
@@ -0,0 +1,171 @@
+specify the complete service definition in a yaml file.
+
+Command format
+--------------
+
+.. prompt:: bash #
+
+   ceph orch apply snmp-gateway <snmp_version:V2c|V3> <destination> [<port:int>] [<engine_id>] [<auth_protocol: MD5|SHA>] [<privacy_protocol:DES|AES>] [<placement>] ...
+
+
+Usage Notes
+
+- you must supply the ``--snmp-version`` parameter
+- the ``--destination`` parameter must be of the format hostname:port (no default)
+- you may omit ``--port``. It defaults to 9464
+- the ``--engine-id`` is a unique identifier for the device (in hex) and required for SNMP v3 only.
+  Suggested value: 8000C53F<fsid> where the fsid is from your cluster, without the '-' symbols
+- for SNMP V3, the ``--auth-protocol`` setting defaults to **SHA**
+- for SNMP V3, with encryption you must define the ``--privacy-protocol``
+- you **must** provide a -i <filename> to pass the secrets/passwords to the orchestrator
+
+Deployment Examples
+===================
+
+SNMP V2c
+--------
+Here's an example for V2c, showing CLI and service based deployments
+
+.. prompt:: bash #
+
+   ceph orch apply snmp-gateway --port 9464 --snmp_version=V2c --destination=192.168.122.73:162 -i ./snmp_creds.yaml
+
+with a credentials file that contains;
+
+.. code-block:: yaml
+
+   ---
+   snmp_community: public
+
+Alternatively, you can create a yaml definition for the gateway and apply it from a single file
+
+.. prompt:: bash #
+
+   ceph orch apply -i snmp-gateway.yml
+
+with the file containing the following configuration
+
+.. code-block:: yaml
+
+    service_type: snmp-gateway
+    service_name: snmp-gateway
+    placement:
+      count: 1
+    spec:
+      credentials:
+        snmp_community: public
+      port: 9464
+      snmp_destination: 192.168.122.73:162
+      snmp_version: V2c
+
+
+SNMP V3 (authNoPriv)
+--------------------
+Deploying an snmp-gateway service supporting SNMP V3 with authentication only, would look like this;
+
+.. prompt:: bash #
+
+   ceph orch apply snmp-gateway --snmp-version=V3 --engine-id=800C53F000000 --destination=192.168.122.1:162 -i ./snmpv3_creds.yml
+
+with a credentials file as;
+
+.. code-block:: yaml
+
+   ---
+   snmp_v3_auth_username: myuser
+   snmp_v3_auth_password: mypassword
+
+or as a service configuration file
+
+.. code-block:: yaml
+
+   service_type: snmp-gateway
+   service_name: snmp-gateway
+   placement:
+     count: 1
+   spec:
+     credentials:
+       snmp_v3_auth_password: mypassword
+       snmp_v3_auth_username: myuser
+     engine_id: 800C53F000000
+     port: 9464
+     snmp_destination: 192.168.122.1:162
+     snmp_version: V3
+
+
+SNMP V3 (authPriv)
+------------------
+
+Defining an SNMP V3 gateway service that implements authentication and privacy (encryption), requires two additional values
+
+.. prompt:: bash #
+
+   ceph orch apply snmp-gateway --snmp-version=V3 --engine-id=800C53F000000 --destination=192.168.122.1:162 --privacy-protocol=AES -i ./snmpv3_creds.yml
```

Comment:
Wanted to try deploying this on some VMs, roughly following the docs here for V3 with authPriv:

```
[ceph: root@vm-00 /]# ceph orch ls --service-name snmp-gateway --format=yaml
service_type: snmp-gateway
service_name: snmp-gateway
placement:
  count: 1
spec:
  credentials:
    snmp_v3_auth_password: mypassword
    snmp_v3_auth_username: myuser
    snmp_v3_priv_password: mysecret
  engine_id: 800C53F000000
  port: 9464
  privacy_protocol: AES
  snmp_destination: 192.168.122.154:162
  snmp_version: V3
status:
  created: '2021-12-15T19:30:24.824432Z'
  last_refresh: '2021-12-15T19:30:40.357717Z'
  ports:
  - 9464
  running: 1
  size: 1
events:
- 2021-12-15T19:30:24.838480Z service:snmp-gateway [INFO] "service was created"
```
It looked like it was created okay (which is great) but then I was seeing error messages spammed in the logs all looking roughly like

```
Dec 15 19:35:48 vm-00 conmon[26279]: time="2021-12-15T19:35:48Z" level=error msg="502 Bad Gateway: err=EngineId must be a hexadecimal string and length is range 5..32, value `800C53F000000` data=&{Receiver:snmp-gateway Status:firing Alerts:[{Status:firing Labels:map[alertname:network packets dropped device:eth0 instance:vm-00 job:node oid:1.3.6.1.4.1.50495.15.1.2.8.2 severity:warning type:ceph_default] Annotations:map[description:Node vm-00 experiences packet drop > 0.01% or > 10 packets/s on interface eth0.\n] StartsAt:2021-12-15 19:29:23.501929571 +0000 UTC EndsAt:0001-01-01 00:00:00 +0000 UTC GeneratorURL:http://vm-00:9095/graph?g0.expr=%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%2F+%28increase%28node_network_receive_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+0.0001+or+%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+10&g0.tab=1 Fingerprint:6e19bb89d1e0a16b} {Status:firing Labels:map[alertname:network packets dropped device:eth0 instance:vm-01 job:node oid:1.3.6.1.4.1.50495.15.1.2.8.2 severity:warning type:ceph_default] Annotations:map[description:Node vm-01 experiences packet drop > 0.01% or > 10 packets/s on interface eth0.\n] StartsAt:2021-12-15 19:29:33.501929571 +0000 UTC EndsAt:0001-01-01 00:00:00 +0000 UTC GeneratorURL:http://vm-00:9095/graph?g0.expr=%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%2F+%28increase%28node_network_receive_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+0.0001+or+%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+10&g0.tab=1 Fingerprint:a8a8ecdc974152d4} {Status:firing Labels:map[alertname:network packets dropped device:eth0 instance:vm-02 job:node oid:1.3.6.1.4.1.50495.15.1.2.8.2 severity:warning type:ceph_default] Annotations:map[description:Node vm-02 experiences packet drop > 0.01% or > 10 packets/s on interface eth0.\n] StartsAt:2021-12-15 19:29:23.501929571 +0000 UTC EndsAt:0001-01-01 00:00:00 +0000 UTC GeneratorURL:http://vm-00:9095/graph?g0.expr=%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%2F+%28increase%28node_network_receive_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_packets_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+0.0001+or+%28increase%28node_network_receive_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29+%2B+increase%28node_network_transmit_drop_total%7Bdevice%21%3D%22lo%22%7D%5B1m%5D%29%29+%3E%3D+10&g0.tab=1 Fingerprint:20f57e0dc7cd8761}] GroupLabels:map[alertname:network packets dropped] CommonLabels:map[alertname:network packets dropped device:eth0 job:node oid:1.3.6.1.4.1.50495.15.1.2.8.2 severity:warning type:ceph_default] CommonAnnotations:map[] ExternalURL:http://vm-00:9093}" source="http_server.go:129"
Dec 15 19:35:48 vm-00 conmon[26279]: 192.168.122.154 - - [15/Dec/2021:19:35:48 +0000] "POST /alerts HTTP/1.1" 502 126
```

The `Bad Gateway:  err=EngineId must be a hexadecimal string and length is range 5..32, value '800C53F000000'` part standing out obviously. I assume I've just done something wrong here but wanted to ask just in case there is some sort of bug.

Reply:
did you set up an snmp receiver running in v3 mode with that user/password and engine_id at 192.168.122.154? It looks to me like an alert fired, but without a receiver I'd expect bad gateway

Reply:
Also, the engine_id would normally be set by the snmp folks for the user. In your case it's not valid because it's not a string that's a multiple of 2 bytes - which I didn't check for. For example, in my local tests using 800C53F0000000 works fine (which is what all the positive tests use in the unit tests IIRC). I'll add a length check to ensure the given engineid is a multiple of 2 in size so we stop this issue at deployment time.